### PR TITLE
Functionality to parse input string into function or variable

### PR DESCRIPTION
**Pull request by cutehedgehog**

### CODE REVIEW

File: Calculator/Calculator/Services/InputPreprocessingService.cs [hunk 1]
```diff
@@ -21,6 +22,114 @@ public string ReplaceUserFunctions(string input)
...
+        if (FunctionValidation(inputFunc))
+        {
+            outputFunc.Name = match.Groups["name"].Value;
+            outputFunc.Variables = new List<string>(match.Groups["variables"].Value.Split(','));
+            outputFunc.Expression = match.Groups["expression"].Value;
+            return outputFunc;
+        }
+        else
+        {
+            throw new ArgumentException("Invalid input string format");
```
Comment (ShamansJackal) [inline Calculator/Calculator/Services/InputPreprocessingService.cs:38]: Invert the if statement and move the exception throwing to the beginning

File: Calculator/Calculator.Tests/Services/InputPreprocessingServiceTests.cs [hunk 2]
```diff
@@ -1,5 +1,157 @@
...
+
+namespace Calculator.Tests.Services;
 
 public class InputPreprocessingServiceTests
 {
+
+	[Fact]
+	public void ProcessFunction_Correct()
+	{
+		var processor = new InputPreprocessingService();
```
Comment (SUPERKISA8BIIT) [inline Calculator/Calculator.Tests/Services/InputPreprocessingServiceTests.cs:11]: Looks like InputPreprocessingService have two responsibilities validation and processing, this logic should be separated

File: Calculator/Calculator/Services/InputPreprocessingService.cs (hunk 1 above)
Comment (SUPERKISA8BIIT) [inline Calculator/Calculator/Services/InputPreprocessingService.cs]: Invert the if statement and move the exception throwing to the beginning

File: Calculator/Calculator/Services/InputPreprocessingService.cs [hunk 3]
```diff
@@ -21,6 +22,114 @@ public string ReplaceUserFunctions(string input)
...
+        }
+    }
+
+	/// <summary>
+	/// Check if input string is a valid function notation
+	/// </summary>
+	/// <param name="inputFunc"></param>
+	/// <returns>True for valid function</returns>
+    public bool FunctionValidation(string inputFunc) 
+    {
```
Comment (SUPERKISA8BIIT) [inline Calculator/Calculator/Services/InputPreprocessingService.cs]: Validation should be moved to separate class

File: Calculator/Calculator/Services/InputPreprocessingService.cs [hunk 4]
```diff
@@ -21,6 +22,114 @@ public string ReplaceUserFunctions(string input)
...
+        Match match = Regex.Match(inputFunc, pattern);
+
+        if (match.Success)
+        {
+            var variables = new List<string>(match.Groups["variables"].Value.Split(','));
+            if (variables.Count == 0)
+            {
+                return false;
+            }
+            else
```
Comment (SUPERKISA8BIIT) [inline Calculator/Calculator/Services/InputPreprocessingService.cs]: else branch may be omitted

File: Calculator/Calculator/Services/InputPreprocessingService.cs [hunk 5]
```diff
@@ -21,6 +22,114 @@ public string ReplaceUserFunctions(string input)
...
+	/// Check if input string is a valid function notation
+	/// </summary>
+	/// <param name="inputFunc"></param>
+	/// <returns>True for valid function</returns>
+    public bool FunctionValidation(string inputFunc) 
+    {
+        string pattern = @"^(?<name>[A-Za-z]\w*)\((?<variables>[\w,]+)\)=(?<expression>.+)$";
+        Match match = Regex.Match(inputFunc, pattern);
+
+        if (match.Success)
```
Comment (SUPERKISA8BIIT) [inline Calculator/Calculator/Services/InputPreprocessingService.cs]: Invert if and then omit else branch

File: Calculator/Calculator/Services/InputPreprocessingService.cs [hunk 6]
```diff
@@ -21,6 +22,114 @@ public string ReplaceUserFunctions(string input)
...
+                    if (variable.Length == 0 || char.IsDigit(variable[0]))
+                    {
+                        return false;
+                    }
+                }
+            }
+
+            List<string> funcBodyVariables = new List<string>();
+            var expr = match.Groups["expression"].Value;
+            //expr = ReplaceUserFunctions(expr);
```
Comment (SUPERKISA8BIIT) [inline Calculator/Calculator/Services/InputPreprocessingService.cs]: Remove unused code

**Pull request merged**